### PR TITLE
Add primary keys in tables lacking them

### PR DESCRIPTION
**Pull request by vincent-olivert-riera**
Some tables from migrate_utils.py and model.py don't have a primary
key. Because of that some RDBMS complain when executing...

  buildbot upgrade-master /path/to/master/directory/

Having primary keys in tables is highly recommended [1], so let's fix
that issue by adding them.

1: https://stackoverflow.com/questions/840162/should-each-and-every-table-have-a-primary-key

Signed-off-by: Vicente Olivert Riera <Vincent.Riera@imgtec.com>

### CODE REVIEW

File: master/buildbot/db/model.py [hunk 1]
```diff
@@ -498,8 +500,10 @@ class Model(base.DBConnectorComponent):
     # the change.
     scheduler_changes = sautils.Table(
         'scheduler_changes', metadata,
-        sa.Column('schedulerid', sa.Integer, sa.ForeignKey('schedulers.id')),
-        sa.Column('changeid', sa.Integer, sa.ForeignKey('changes.changeid')),
+        sa.Column('schedulerid', sa.Integer, sa.ForeignKey('schedulers.id'),
+                  primary_key=True),
+        sa.Column('changeid', sa.Integer, sa.ForeignKey('changes.changeid'),
+                  primary_key=True),
```
Comment (seankelly) [inline master/buildbot/db/model.py]: Does this create a composite key or try to create two primary keys?

Comment (vriera) [inline master/buildbot/db/model.py]: That's how a composite key is defined using SQLAlchemy.

Comment (tardyp) [inline master/buildbot/db/model.py]: thats pretty cool

File: master/buildbot/db/migrate/versions/053_add_change_properties_changepropertyid.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import sqlalchemy as sa
+
+from buildbot.util import sautils
+
+
+def upgrade(migrate_engine):
+    metadata = sa.MetaData()
+    metadata.bind = migrate_engine
+    change_properties_table = sautils.Table('change_properties', metadata, autoload=True)
+    changepropertyid = sa.Column('changepropertyid', sa.Integer, nullable=True)
```
Comment (tardyp) [inline master/buildbot/db/migrate/versions/053_add_change_properties_changepropertyid.py]: you can just create one migration script which adds the three keys.

then you will only need one unit test file.
The unit test will just make sure that the column is really created.

**Comment by tardyp** [comment]
hi @vriera . Do you intend to finish this?

**Comment by Frodox** [comment]
ping @vincent-olivert-riera

**Comment by vincent-olivert-riera** [comment]
I'm not out from GitHub, I just don't have enough time to look at this. The only part that is missing is the one with the tests. The changes to the model work fine.

**Comment by tardyp** [comment]
@vincent-olivert-riera thanks for dropping by.
We are still interested by those changes. The unit tests are as important as "real code" for us, as our only way to keep the code maintainable.

Hopefully you can find time to finish your upstreaming task. I hope your manager can understand it is important for your team to avoid forking open source projects :)

**Comment by rjarry** [comment]
Maybe I can take this over and include it along with #3815 

@vincent-olivert-riera is that OK with you?

**Comment by vincent-olivert-riera** [comment]
> Maybe I can take this over and include it along with #3815
> @vincent-olivert-riera is that OK with you?

Yes, absolutely. Go for it :-)

**Comment by rjarry** [comment]
@vincent-olivert-riera wrote:
> Yes, absolutely. Go for it :-)

Ok,  I am having a closer look.

First question: why did you add autoincrement primary keys for some tables and sticked with composite keys for others? Shouldn't we keep the schema as close as it is for now and only add composite keys where necessary?

**Comment by rjarry** [comment]
What do you think about the following modification:

```diff
diff --git a/master/buildbot/db/model.py b/master/buildbot/db/model.py
index 5924a5286ebb..8e9a87174349 100644
--- a/master/buildbot/db/model.py
+++ b/master/buildbot/db/model.py
@@ -116,11 +116,9 @@ class Model(base.DBConnectorComponent):
     # This table contains the build properties
     build_properties = sautils.Table(
         'build_properties', metadata,
-        sa.Column('buildpropertyid', sa.Integer, primary_key=True,
-                  autoincrement=True),
         sa.Column('buildid', sa.Integer, sa.ForeignKey('builds.id'),
-                  nullable=False),
-        sa.Column('name', sa.String(256), nullable=False),
+                  primary_key=True),
+        sa.Column('name', sa.String(256), primary_key=True),
         # JSON encoded value
         sa.Column('value', sa.Text, nullable=False),
         sa.Column('source', sa.String(256), nullable=False),
@@ -204,11 +202,9 @@ class Model(base.DBConnectorComponent):
     # This table contains input properties for buildsets
     buildset_properties = sautils.Table(
         'buildset_properties', metadata,
-        sa.Column('buildsetpropertyid', sa.Integer, primary_key=True,
-                  autoincrement=True),
         sa.Column('buildsetid', sa.Integer, sa.ForeignKey('buildsets.id'),
-                  nullable=False),
-        sa.Column('property_name', sa.String(256), nullable=False),
+                  primary_key=True),
+        sa.Column('property_name', sa.String(256), primary_key=True),
         # JSON-encoded tuple of (value, source)
         sa.Column('property_value', sa.Text, nullable=False),
     )
@@ -309,20 +305,17 @@ class Model(base.DBConnectorComponent):
     # Files touched in changes
     change_files = sautils.Table(
         'change_files', metadata,
-        sa.Column('changefileid', sa.Integer, primary_key=True, autoincrement=True),
         sa.Column('changeid', sa.Integer, sa.ForeignKey('changes.changeid'),
-                  nullable=False),
-        sa.Column('filename', sa.String(1024), nullable=False),
+                  primary_key=True),
+        sa.Column('filename', sa.String(1024), primary_key=True),
     )
 
     # Properties for changes
     change_properties = sautils.Table(
         'change_properties', metadata,
-        sa.Column('changepropertyid', sa.Integer, primary_key=True,
-                  autoincrement=True),
         sa.Column('changeid', sa.Integer, sa.ForeignKey('changes.changeid'),
-                  nullable=False),
-        sa.Column('property_name', sa.String(256), nullable=False),
+                  primary_key=True),
+        sa.Column('property_name', sa.String(256), primary_key=True),
         # JSON-encoded tuple of (value, source)
         sa.Column('property_value', sa.Text, nullable=False),
     )
```

**Pull request closed without merging**